### PR TITLE
fix: fork logic

### PR DESCRIPTION
**Pull request by zhangsoledad**
fix issue: `get_forks` may push genesis in old_blocks, cause tx_pool infinite loop

### CODE REVIEW

File: chain/src/chain.rs [hunk 1]
```diff
@@ -280,74 +289,86 @@ impl<CI: ChainIndex + 'static> ChainService<CI> {
...
+        let mut is_new_record = true;
 
-        // The old fork may longer than new fork
-        if number < current_tip_number {
-            for n in number..=current_tip_number {
-                let hash = self.shared.block_hash(n).unwrap();
+        // new_blocks = forks[latest_common + 1 .. new_tip]
+        // old_blocks = chain[latest_common + 1 .. old_tip]
+        new_blocks.push(new_tip_block.clone());
+        let mut index = new_tip_number - 1;
```
Comment (jjyr) [inline chain/src/chain.rs]: I believe `common_parent_number` is more accurate for this variable and can help a lot for others who read this code.

File: chain/src/chain.rs [hunk 2]
```diff
@@ -280,74 +289,86 @@ impl<CI: ChainIndex + 'static> ChainService<CI> {
...
-                if old_hash == hash {
-                    break;
-                }
-
-                let old_block = self.shared.block(&old_hash).unwrap();
-                old_blocks.push(old_block);
+            if index == 0 {
+                break;
+            }
+            let old_hash = self.shared.block_hash(index).unwrap();
```
Comment (jjyr) [inline chain/src/chain.rs]: `common_parent_hash` or something like it is a more related name.

File: chain/src/chain.rs (hunk 1 above)
Comment (jjyr) [reply chain/src/chain.rs]: or common_block_number

File: chain/src/chain.rs [hunk 3]
```diff
@@ -280,74 +289,86 @@ impl<CI: ChainIndex + 'static> ChainService<CI> {
...
 
-        // The old fork may longer than new fork
-        if number < current_tip_number {
-            for n in number..=current_tip_number {
-                let hash = self.shared.block_hash(n).unwrap();
+        // new_blocks = forks[latest_common + 1 .. new_tip]
+        // old_blocks = chain[latest_common + 1 .. old_tip]
+        new_blocks.push(new_tip_block.clone());
+        let mut index = new_tip_number - 1;
+        let mut index_hash = new_tip_block.header().parent_hash().clone();
```
Comment (jjyr) [inline chain/src/chain.rs]: Use some name like `new_chain_index` to distinguish it with the `index` variable, those two value is confused.

File: chain/src/chain.rs [hunk 4]
```diff
@@ -280,74 +289,86 @@ impl<CI: ChainIndex + 'static> ChainService<CI> {
...
 
-            if is_open {
-                let ext = self.shared.block_ext(&hash).unwrap();
+            let new_block = self.shared.block(&index_hash).unwrap();
+            index_hash = new_block.header().parent_hash().clone();
+            new_blocks.push(new_block);
+
+            if is_new_record {
+                let ext = self.shared.block_ext(&index_hash).unwrap();
                 if ext.valid.is_none() {
```
Comment (jjyr) [inline chain/src/chain.rs]: I still can't get it, what is the intention here?

File: rpc/src/module/miner.rs [hunk 5]
```diff
@@ -49,23 +51,35 @@ impl<CI: ChainIndex + 'static> MinerRpc for MinerRpcImpl<CI> {
...
+                Ok(Some(block.header().hash().clone()))
+            } else {
+                debug!(target: "rpc", "submit_block process_block {:?}", ret);
+                Ok(None)
+            }
         } else {
-            debug!(target: "rpc", "submit_block process_block {:?}", ret);
-            Err(Error::internal_error())
+            debug!(target: "rpc", "submit_block header verifier {:?}", header_verify_ret);
+            Ok(None)
```
Comment (jjyr) [inline rpc/src/module/miner.rs]: Why return an Option?

File: chain/src/chain.rs (hunk 2 above)
Comment (zhangsoledad) [reply chain/src/chain.rs]: It's NOT `common_parent_hash`

File: chain/src/chain.rs (hunk 1 above)
Comment (zhangsoledad) [reply chain/src/chain.rs]: np, `common_block_number` `common_parent_number` make no sense.
it's just an index of an enumerator.
check https://github.com/bitcoin/bitcoin/blob/v0.17.1/src/chain.cpp#L51. almost like this.

File: rpc/src/module/miner.rs (hunk 5 above)
Comment (zhangsoledad) [reply rpc/src/module/miner.rs]: It's NOT an `internal_error`,  submit_block SHOULD provide share rejection reasons.
but this needs much further work, temporarily return an Option.

this patch fixes We haven't verified the block header

File: chain/src/chain.rs (hunk 1 above)
Comment (jjyr) [reply chain/src/chain.rs]: It is ok to use a variable like `index` in a short code snippet, but for here it confuses.

Moreover, we have two confused variables, at least we need a name to be enough to distinguish the two variables usage. 

I think `new_chain_index` and `block_number`, or other names definitely could be great than `index` and `index_hash`.

File: chain/src/chain.rs (hunk 4 above)
Comment (zhangsoledad) [reply chain/src/chain.rs]: It's an unseen mark, maybe name it unseen directly.

Comment (zhangsoledad) [reply chain/src/chain.rs]: btw. It's a hotfix, not a refactory,

**Comment by zhangsoledad** [comment]
cc @jjyr , have some updates.

**Comment by jjyr** [comment]
The `chain` module should be refactoring in further.

**Pull request merged**